crates/core/src/models/call.rs

```diff
@@ -0,0 +1,328 @@
+use std::slice::SliceIndex;
+
+use reth_primitives::TransactionSigned;
+use reth_rlp::Decodable;
+use starknet::accounts::Call;
+use starknet_crypto::FieldElement;
+
+use super::ConversionError;
+use crate::models::DataDecodingError;
+
+pub struct Calls(Vec<Call>);
+
+impl From<Calls> for Vec<Call> {
+    fn from(calls: Calls) -> Self {
+        calls.0
+    }
+}
+
+impl From<Vec<Call>> for Calls {
+    fn from(calls: Vec<Call>) -> Self {
+        Self(calls)
+    }
+}
+
+/// TryFrom implementation for account contract calls
+impl TryFrom<Vec<FieldElement>> for Calls {
+    type Error = ConversionError;
+
+    fn try_from(value: Vec<FieldElement>) -> Result<Self, Self::Error> {
+        let calls_len = u32::try_from(value[0]).map_err(|e| ConversionError::ValueOutOfRange(e.to_string()))? as usize;
+        let mut offset = calls_len * 4 + 2;
+
+        let mut calls = vec![];
+        for i in 0..calls_len {
+            let calldata_len =
+                u32::try_from(value[i * 4 + 4]).map_err(|e| ConversionError::ValueOutOfRange(e.to_string()))? as usize;
+            let call = Call {
+                to: value[i * 4 + 2],
+                selector: value[i * 4 + 3],
+                calldata: value[offset..offset + calldata_len].to_vec(),
+            };
+            offset += calldata_len;
+            calls.push(call);
+        }
+        Ok(Calls(calls))
+    }
+}
+
+impl TryFrom<&Calls> for TransactionSigned {
+    type Error = DataDecodingError;
+
+    fn try_from(value: &Calls) -> std::result::Result<Self, Self::Error> {
+        let call = value.0[0] // for now we decode signature only from the first call
+            .calldata
+            .iter()
+            .filter_map(|x| u8::try_from(*x).ok())
+            .collect::<Vec<u8>>();
+        TransactionSigned::decode(&mut call.as_slice())
+            .map_err(|e| DataDecodingError::SignatureDecodingError(e.to_string()))
+    }
+}
+
+impl Calls {
+    pub fn len(&self) -> usize {
+        self.0.len()
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+
+    pub fn get<I>(&self, index: I) -> Option<&I::Output>
+    where
+        I: SliceIndex<[Call]>,
+    {
+        self.0.get(index)
+    }
+
+    pub fn get_eth_transaction(&self) -> Result<TransactionSigned, DataDecodingError> {
+        self.try_into()
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::str::FromStr;
+
+    use reth_primitives::{Address, U256};
+
+    use super::*;
+
+    fn to_vec_field_element(vec: Vec<&str>) -> Vec<FieldElement> {
+        vec.into_iter().filter_map(|f| FieldElement::from_hex_be(f).ok()).collect()
+    }
+
+    fn get_simple_test_calldata() -> Vec<FieldElement> {
+        let calldata = vec![
```

Comment:
can't we generate a raw tx with starknet rs?

Reply:
I don't think so, can't find anywhere in starknet rs how I could do this, and doesn't seem to be the case in [tests](https://github.com/xJonathanLEI/starknet-rs/blob/b9e3857fe89b58ca8b24e95d3adfd85c1f8a3d04/starknet-providers/tests/jsonrpc.rs#L288).
I can however move this to a json file if you find it cleaner.